modules/antitracking/sources/attrack.es

```diff
@@ -244,507 +245,517 @@ export default class CliqzAttrack {
-              annotations.counter = annotations.counter || new TrackerCounter();
-              annotations.counter.addTrackerSeen(state.ghosteryBug, state.urlParts.hostname);
-            }
-            if (state.ghosteryBug && this.config.cookieMode === COOKIE_MODE.GHOSTERY) {
-              // track domains used by ghostery rules so that we only block cookies for these
-              // domains
-              this.ghosteryDomains[state.urlParts.generalDomain] = state.ghosteryBug;
-            }
-          },
-        },
-        {
-          name: 'checkExternalBlocking',
-          spec: 'blocking',
-          fn: (state, response) => {
-            if (response.cancel === true || response.redirectUrl) {
-              state.incrementStat('blocked_external');
-              response.shouldIncrementCounter = true;
-              return false;
-            }
-            return true;
-          },
-        },
-        {
-          name: 'checkDatabaseEnabled',
-          spec: 'break',
-          // if there is no database the following steps should be skipped
-          fn: () => this.config.databaseEnabled,
-        },
-        {
-          name: 'tokenExaminer.examineTokens',
-          spec: 'collect', // TODO - global state
-          fn: state => steps.tokenExaminer.examineTokens(state),
-        },
-        {
-          name: 'tokenTelemetry.extractKeyTokens',
-          spec: 'collect', // TODO - global state
-          fn: state => !steps.tokenTelemetry || steps.tokenTelemetry.extractKeyTokens(state),
-        },
-        {
-          name: 'tokenChecker.findBadTokens',
-          spec: 'annotate',
-          fn: state => steps.tokenChecker.findBadTokens(state),
-        },
-        {
-          name: 'checkSourceWhitelisted',
-          spec: 'break',
-          fn: (state) => {
-            if (this.checkIsWhitelisted(state)) {
-              state.incrementStat('source_whitelisted');
-              return false;
-            }
-            return true;
-          },
-        },
-        {
-          name: 'checkShouldBlock',
-          spec: 'break',
-          fn: state => state.badTokens.length > 0 && this.qs_whitelist.isUpToDate()
-                       && !this.config.paused,
-        },
-        {
-          name: 'oauthDetector.checkIsOAuth',
-          spec: 'break',
-          fn: state => steps.oauthDetector.checkIsOAuth(state, 'token'),
-        },
-        {
-          name: 'isQSEnabled',
-          spec: 'break',
-          fn: () => this.isQSEnabled(),
-        },
-        {
-          name: 'blockRules.applyBlockRules',
-          spec: 'blocking',
-          fn: (state, response) => steps.blockRules.applyBlockRules(state, response),
-        },
-        {
-          name: 'logBlockedToken',
-          spec: 'collect',
-          fn: (state) => {
+    // initialise step objects
+    const pendingInits = [];
+    for (const key of Object.keys(steps)) {
+      const step = steps[key];
+      if (step.init) {
```

Comment:
Note: step.init may return a promise, but not always. That means, something like "step.init().catch(...)" will break.